### ISSUE TITLE
02-sgd/src/sgd.py gradients for pixels are the same

### ISSUE THREAD

**Issue by smao-astro** [MPoL-dev/examples#2]
Hi,

Thank you for open-source this tutorial. I am learning RML with 02-sgd/src/sgd.py. I start with command

`src/sgd.py --tensorboard-log-dir=runs/sgd-debug-ent0 --save-checkpoint=checkpoints/sgd-debug-ent0.pt --lr=1 --FWHM=0.05 --epochs=10 --lam-ent=1e-5`

and I found **my image cube stuck at a flat image with constant pixel values in training**.

To debug, I add code below after line https://github.com/MPoL-dev/examples/blob/eec19e5215487008e77d5281fa6837f7b82ed944/02-sgd/src/sgd.py#L75

```
    # Analyze data range and percentiles
    min_val = np.min(img)
    max_val = np.max(img)
    p05 = np.percentile(img, 5)
    p95 = np.percentile(img, 95)
    print(f"Image stats at step {step}:")
    print(f"  Range: [{min_val:.3e}, {max_val:.3e}]")
    print(f"  5-95%: [{p05:.3e}, {p95:.3e}]")
    
    # Check if all pixels have the same value
    std_val = np.std(img)
    unique_vals = np.unique(img)
    print(f"  Standard deviation: {std_val:.3e}")
    print(f"  Number of unique values: {len(unique_vals)}")
    if len(unique_vals) <= 5:  # Only print if there are few unique values
        print(f"  Unique values: {unique_vals}")
    
    if std_val < 1e-10:  # Threshold for considering values equal
        print("  WARNING: All pixels appear to have the same value!")
    
    # Print gradient statistics
    print("\nGradient statistics:")
    for name, param in model.named_parameters():
        if param.grad is not None:
            grad = param.grad.detach().cpu().numpy()
            print(f"  {name}:")
            print(f"    Mean: {np.mean(grad):.3e}")
            print(f"    Std: {np.std(grad):.3e}")
            print(f"    Min: {np.min(grad):.3e}")
            print(f"    Max: {np.max(grad):.3e}")
            print(f"    Norm: {np.linalg.norm(grad):.3e}")
            
            # For base_cube, analyze individual pixel gradients
            if name == "bcube.base_cube":
                print("\n  Base cube pixel analysis:")
                # Reshape to 2D for easier analysis
                grad_2d = grad[0]  # Take first channel
                param_2d = param.detach().cpu().numpy()[0]  # Take first channel
                
                # Sample some random pixels
                npix = grad_2d.shape[0]
                sample_size = min(5, npix)
                rows = np.random.choice(npix, sample_size, replace=False)
                cols = np.random.choice(npix, sample_size, replace=False)
                
                print("    Random pixel samples (row, col, value, gradient):")
                for r, c in zip(rows, cols):
                    print(f"      ({r}, {c}): value={param_2d[r,c]:.3e}, grad={grad_2d[r,c]:.3e}")
                
                # Check if all gradients are identical
                if np.allclose(grad_2d, grad_2d[0,0], rtol=1e-10):
                    print("    WARNING: All gradients are identical!")
                else:
                    print(f"    Gradient variation: {np.std(grad_2d):.3e}")
                
                # Check if all parameter values are identical
                if np.allclose(param_2d, param_2d[0,0], rtol=1e-10):
                    print("    WARNING: All parameter values are identical!")
                else:
                    print(f"    Parameter variation: {np.std(param_2d):.3e}")
```

I got outputs confirms my hypothesis: **the gradient for each pixel are identical** for some reason I don't get

> Gradient statistics:
>   bcube.base_cube:
>     Mean: 6.661e-09
>     Std: 1.021e-14
>     Min: 6.661e-09
>     Max: 6.661e-09
>     Norm: 6.848e-06
> 
>   Base cube pixel analysis:
>     Random pixel samples (row, col, value, gradient):
>       (704, 369): value=-8.994e+00, grad=6.661e-09
>       (230, 599): value=-8.994e+00, grad=6.661e-09
>       (103, 187): value=-8.994e+00, grad=6.661e-09
>       (728, 596): value=-8.994e+00, grad=6.661e-09
>       (352, 609): value=-8.994e+00, grad=6.661e-09
>     WARNING: All gradients are identical!
>     WARNING: All parameter values are identical!
> Train Epoch: 0 [40000/96266 (41%)]      Loss: 0.399863
> Image stats at step 20:
>   Range: [1.119e-04, 1.119e-04]
>   5-95%: [1.119e-04, 1.119e-04]
>   Standard deviation: 6.548e-11
>   Number of unique values: 1
>   Unique values: [0.00011193]
>   WARNING: All pixels appear to have the same value!

May I ask if this is by design, or some unwanted issues?

Thanks,
Shunyuan

**Comment by smao-astro** [comment]
okay I resolve the issue by initializing the cube with random non-uniform noise on a uniform background

**Comment by smao-astro** [comment]
Hi @iancze 
 
I found for examples/02-sgd/sgd.py

1. if optimize both negative log likelihood and entropy (`lam-ent` non-zero), the gradients are non-zeros
2. if optimize only negative log likelihood (`lam-ent` zero), the gradients are zeros

To show this behavior, I add code below to train() in sgd.py

```
        # Print gradient statistics
        print("\nGradient statistics:")
        for name, param in model.named_parameters():
            if param.grad is not None:
                grad = param.grad.detach().cpu().numpy()
                print(f"  {name}:")
                print(f"    Mean: {np.mean(grad):.3e}")
                print(f"    Std: {np.std(grad):.3e}")
                print(f"    Min: {np.min(grad):.3e}")
                print(f"    Max: {np.max(grad):.3e}")
                print(f"    Norm: {np.linalg.norm(grad):.3e}")

```

With `lam-ent=0`, I got 

```
Gradient statistics:
  bcube.base_cube:
    Mean: 0.000e+00
    Std: 0.000e+00
    Min: -0.000e+00
    Max: 0.000e+00
    Norm: 0.000e+00
Train Epoch: 5 [0/96266 (0%)]   Loss: 19.303207
```

I am not expecting this. The nll loss is non-zero, so its gradient back-propagate to base cube pixels should be non-zero.

Please have a look at it and help me to figure out the reason of it.

Thanks.

**Comment by smao-astro** [comment]
I found the issue only occur for Apple GPU (mps), fall back to cpu resolve the issue.
